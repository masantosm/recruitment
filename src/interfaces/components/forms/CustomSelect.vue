<template>
  <div class="w-full relative" data-test-id="custom-select">
    <div
      data-test-id="select"
      class="flex justify-between flex-1 border border-border p-1 rounded-md"
      @click="toggleDropdown"
    >
      <span class="text-sm pl-2 text-center">{{ selectedItem || 'Selected item' }}</span>
      <span>&#9660;</span>
    </div>
    <div v-if="isOpen" class="w-full absolute bg-white border border-border rounded-lg shadow-md">
      <div
        :data-test-id="`option-${index}`"
        class="p-2 hover:bg-light-grey cursor-pointer z-10"
        v-for="(item,index) in options"
        :key="index"
        @click="selectItem(item)"
      >
        {{ item.name }}
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { defineComponent, ref } from 'vue'
import type { Status } from '../../../domain/models/Status'

export default defineComponent({
  name: 'CustomSelect',
  props: {
    options: {
      type: Array as () => Status[],
      required: true
    },
    modelValue: {
      type: String,
      default: null
    }
  },
  emits: ['update:modelValue'],
  setup(props, { emit }) {
    const selectedItem = ref<string | null>(
      props.options?.find((elem) => elem.id === props.modelValue)?.name || null
    )
    const isOpen = ref<boolean>(false)

    const toggleDropdown = () => {
      isOpen.value = !isOpen.value
    }

    const selectItem = (item: Status) => {
      selectedItem.value = item.name
      isOpen.value = false
      emit('update:modelValue', item.id)
    }

    return {
      isOpen,
      selectedItem,
      toggleDropdown,
      selectItem
    }
  }
})
</script>

<style scoped>
</style>
