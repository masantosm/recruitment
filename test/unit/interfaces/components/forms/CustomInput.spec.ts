import { shallowMount } from '@vue/test-utils'
import { describe, it, expect } from 'vitest'
import CustomSelect from '../../../../../src/interfaces/components/forms/CustomInput.vue'

describe('CustomInput component is definied', () => {
  it('renders correctly with the given props', () => {
    const wrapper = shallowMount(CustomSelect, {
      global: {},
      props: {
        label: 'name',
        modelValue: 'Test'
      }
    })
    const inputComponent = wrapper.find('[data-test-id="custom-input-name"]')
    expect(inputComponent.exists()).toBeTruthy()
    const inputLabel = wrapper.find('[data-test-id="input-label"]')
    expect(inputLabel.text()).toContain('name')
    const input = wrapper.find('[data-test-id=input]')
    expect(input.attributes('value')).toBe('Test')
    input.trigger('input')
    const emittedEvent = wrapper.emitted('update:modelValue')
    expect(emittedEvent).toBeTruthy()
    if (emittedEvent) {
      expect(emittedEvent[0][0]).toBe('Test')
    }
  })
})
