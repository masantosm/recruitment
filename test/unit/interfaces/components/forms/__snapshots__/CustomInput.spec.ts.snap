// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CustomSelect.vue > matches snapshot 1`] = `
"<div class="w-full relative" label="Username">
  <div class="flex justify-between flex-1 border border-border p-1 rounded-md"><span class="text-sm pl-2 text-center">Selected item</span><span>â–¼</span></div>
  <!--v-if-->
</div>"
`;
